### DISCUSSION TITLE
Date / Time formatting

### DISCUSSION THREAD

**Discussion by jfrancais** [Blackymas/NSPanel_HA_Blueprint#523] [Q&A]
I've been tinkering with the date/time format in the blueprint.  I've been able to create an input field for both the date format and the time format so I can change them from within the blueprint input upon setup.  I did notice a rather large section of the blueprint dedicated to translating the day and month name for the selected language.  It made me wonder, if you used %b (Month) and %a (day) in timestamp_custom() doesn't that handle the translation for you using the Home Assistant configured language?  It would save a lot of code space to be able to adjust the date##_mui assignments using one format for all.

In my case, I changed line 3579:
{%- elif language == "ENG" -%} {{ ((as_timestamp(now())) | timestamp_custom('%a %d.%m' )) | replace('Mon', 'Monday,')| replace('Tue', 'Tuesday,')| replace('Wed', 'Wednesday,')| replace('Thu', 'Thursday,')| replace('Fri', 'Friday,')| replace('Sat', 'Saturday,')| replace('Sun', 'Sunday,') }}

to
    {%- elif language == "ENG" -%} {{ ((as_timestamp(now())) | timestamp_custom(nspanel_date_format ))  }}

where nspanel_date_format is an input from the blueprint set to "%a %b %-d" and it now displays
Fri Mar 24

If the language was localized, all those language if blocks could be set to the same format that could be adjusted from an input.

**Comment by jfrancais** [comment]
![blueprint](https://user-images.githubusercontent.com/18350834/227636743-5eb40188-371e-49c5-9894-642ae9acc375.jpg)

**Reply by edwardtfn** [reply]
This is a possibility, but it requires some knowledge from users about those codes, which many times is not the case and, in the end, would be another point for questions every time some user gets an issue with the display.

Have you tried the Blueprint from dev? There are pre-defined options for date & time formats already. Do you need something different than that?
![IMG_6116](https://user-images.githubusercontent.com/94725493/227729501-bf2ee05c-60cf-45a8-9601-70d8e54fcf3b.png)

**Reply by jfrancais** [reply]
I haven't.  I just noticed that today after someone suggested I do a pull request for my changes.  I don't have much GitHub experience, but do have a bunch of experience doing code work (I was a CVS guy back when I used to code). I like having the custom format ability but having it default to the standard would help those unfamiliar with the options.  What I'm really curious about it the language.  If the month and day name are translated to local language just by using the formatting codes it definitely makes the blueprint code more concise.

**Reply by edwardtfn** [reply]
> If the month and day name are translated to local language just by using the formatting codes it definitely makes the blueprint code more concise.

I agree that would be a more concise approach, however, there are some points:
- As far as I know, Home Assistant have language selection per user, but automations (including the ones based on Blueprints) are executed by the system which will always be in English (I'm not sure which one, would have to check the docs), so that wouldn't consider the user selection.
We probably could still use the locale code to force the format to a specific time/date format, but that would require changing the language selection, as en-US and en-UK will have different time selection, so this will force to have more options on the language selection and a more complex translation for the other strings not relate to date/time, which would have other benefits, but increase complexity.
- The panel is very limited in space and memory, so having some control of what we send will reduce the need of supporting users on all the different possibilities they could try.
Also, we could play a bit more when using pre-defined formats. If you take a look on the version in DEV right now, we have created a different elememt on the panel to have the AM/PM info splitter from the time string, so we could reduce the font a bit ij order to fit in the limited space available in the screen.

And in the end you can still play a bit with those codes for the by editing the date and time files on the automation on `automations.yaml`.

By the way, how would you like to see the date and time displayed?

**Reply by jfrancais** [reply]
Looked at doing a pull request but I don't think I know what I'm doing on that front.  

            - label: '%B %-d (ex January 1)' 
              value: '%B %-d'  

That would be the date formatting I added to my yaml file to get the date format I use.

I have several other tweaks I have incorporated into my yaml.  Is there a good tutorial you can point me to regarding pull requests and how to present these for review?

**Comment by jpettitt** [comment]
Is there a PR with your changes?  I'd like to try them out.

**Reply by jfrancais** [reply]
I don't but I'll investigate doing that this weekend.

**Comment by rleongcs** [comment]
@edwardtfn Any possibility to add the format like what @jfrancais was requesting initially?

I was looking at:
- Wed, 19 Apr
- Wed, Apr 19
- Wednesday, 19 Apr (Not sure if enough screen estate)
- Wednesday, Apr 19 (Not sure if enough screen estate)

Thanks.

**Reply by edwardtfn** [reply]
Everything is possible, but as I mentioned on [this post](https://github.com/Blackymas/NSPanel_HA_Blueprint/discussions/523#discussioncomment-5428536), this can requires more effort than it sounds initially.
About the options you are proposing:
1. `Wed, 19 Apr` and `Wed, Apr 19`: I believe those are going to fit fine in the display but it would require translations for all the languages for the short week day name (`Mon`, `Tue`, etc.) and for the short months names (`Jan`, `Feb`, etc.) and in the end we have to make sure all the supported languages can provide a shorter version which fits into the space available in the screen. But again, I think this is feasible.
1. `Wednesday, 19 Apr` and `Wednesday, Apr 19`: The EU version is limited to something around 13-chars on this string and the US version is not much better than this, so I can anticipate this one will be exceeding the space available, specially when we expand this to other languages (in Portuguese this would be `Segunda-feira, 19 Abr`, which is 21-chars long). Once 1 is implemented, this one would be quite easy from the programming side, as week days are already translated, but it would be a challenge from the UI perspective.
1. None of those options are exactly the same as the one proposed by @jfrancais (`Fri Mar 24`).

**Reply by rleongcs** [reply]
Sure...This is not a high priority. Whenever you guys are able to implement it and not having too much of a trouble.

`None of those options are exactly the same as the one proposed by @jfrancais (Fri Mar 24).`
I would say the first option is close and usally there is a comma in the format.

**Reply by rleongcs** [reply]
Just made some changes to the blueprint yaml file while waiting for the official 😄 

Ended up incorporating the year as there are lots of space once the Day of Week and Month are abbreviated.

![IMG_1585](https://user-images.githubusercontent.com/61628312/234622348-d5b2aa50-5ff6-4aa4-bdd5-12cb995c9bc2.jpg)

**Comment by denes44** [comment]
Maybe a custom option would be enough for advanced users, where they can enter a custom formatting string. That way everybody can select one of the four options, or if it's not fit for them, they can research how they can use a custom string.
For example I like to use the month/date format, but with leading zeros.

**Comment by rleongcs** [comment]
Or maybe even an "Advanced" option with many other things thrown in there besides this date format. For example like #730

**Comment by denes44** [comment]
If anybody wants to help me with the custom date/time formatting, I started doing it, I added an option in the blueprint UI to select and specify a custom format:
![nspanel_datetime](https://user-images.githubusercontent.com/60078357/234862395-de2dde50-e65a-42f3-a31f-b6792164f982.jpg)
But I'm having trouble making the templating part work. Have a look at [this commit](https://github.com/Blackymas/NSPanel_HA_Blueprint/commit/bf200fe37994824603fef202f475fcb21b0ca9db).

**Comment by edwardtfn** [comment]
This is my proposal to solve this for now: #744 

Please take a look and let me know your thoughts.

**Reply by denes44** [reply]
Well, editing the format string in YAML mode is easy enough, that's true.
However I still think it's better if it is shown in the UI. Both the fact that a custom option is selected, and the custom format string as well.

It is easier for the users as well to discover this feature, if they see a custom option in the UI. Otherwise they may not realize they can do it, and will ask here every time, how it can be changed.

**Reply by edwardtfn** [reply]
> It is easier for the users as well to discover this feature, if they see a custom option in the UI. Otherwise they may not realize they can do it, and will ask here every time, how it can be changed

This can be handled by good documentation. ;)

**Reply by edwardtfn** [reply]
> However I still think it's better if it is shown in the UI. Both the fact that a custom option is selected, and the custom format string as well.

My point with this is that the custom format can be easy for people with some programing skills, but is very confusing for everyone else. So that would require really good documentation on how to use that and also explaining that it would be used only if is selected 'custom' in the other field. And we will have also to explain that if they use `%a`, `%b` or `%B` the string will be shown in English and not in the language they selected...
If my mom cannot understand the feature, it is "advanced". 😉
I think  in the end this will be for advanced users only, so YAML editing should be OK.
If we want this for non advanced users I will rather then add all (or as many as possible) possibilities in the selector, so they can choose instead of typing programing codes. I don't see any problem about having dozens of options there.
I don't see as a big issue also translating the remaining strings... It's not that much.

**Reply by edwardtfn** [reply]
And now I've added support for translations of all date related strings:
![IMG_3336](https://user-images.githubusercontent.com/94725493/235082477-2a26c7c7-424b-478c-9a17-41e3bf5994a8.jpeg)
![IMG_3351](https://user-images.githubusercontent.com/94725493/235082496-14bb7e74-8ea6-41cc-8ac8-708ac2a22f04.jpeg)
![IMG_3352](https://user-images.githubusercontent.com/94725493/235082504-322c4075-4de2-4ce8-b218-3025ada25ec4.jpeg)
![IMG_3353](https://user-images.githubusercontent.com/94725493/235082509-60c3adc3-5a65-44f0-af72-5591831abd56.jpeg)
![IMG_3354](https://user-images.githubusercontent.com/94725493/235082515-c2215f24-fc99-47f2-bdfb-2a6a94465e45.jpeg)
![IMG_3355](https://user-images.githubusercontent.com/94725493/235082519-7f159733-e372-4ea5-b305-fd6fe86cc1c2.jpeg)
![IMG_3356](https://user-images.githubusercontent.com/94725493/235082528-347fc33e-d3ee-468c-9273-19227dc2fcd7.jpeg)
![IMG_3357](https://user-images.githubusercontent.com/94725493/235082534-ed2cb4db-edd0-4f90-9caa-aaf07aeb905b.jpeg)
![IMG_3358](https://user-images.githubusercontent.com/94725493/235082537-2ffead80-a1f9-432b-886c-325d3d20b4d4.jpeg)

**Reply by edwardtfn** [reply]
And the support for custom formats:
![image](https://user-images.githubusercontent.com/94725493/235084977-f8dc87cb-3073-41a5-bea8-196d3c15e64e.png)

**Reply by rleongcs** [reply]
Just to confirm...do you have to edit the yaml or just enter the custom format in the UI? As i saw from another thread talking about editing the yaml.

**Reply by edwardtfn** [reply]
With this last commit you can type custom format directly in the UI, so no need to play with yaml.

**Reply by rleongcs** [reply]
Wow...very nice. Thank you.

**Reply by rleongcs** [reply]
Tested the custom date format and it is working great in Dev branch. 🎉

**Comment by rleongcs** [comment]
I could see the US version has a slight edge over the EU version as it has more exposed pixels. You can actually have a very long date format unlike the EU version where the weather info will cause issue. Too bad the form factor is not gonna fit my wall. 😞

**Reply by edwardtfn** [reply]
That's right...
I've compared those versions on this post where I mentioned about the additional pixels available: https://github.com/Blackymas/NSPanel_HA_Blueprint/discussions/713

**Reply by rleongcs** [reply]
Yes...I have read that post.

**Comment by Zaleo80** [comment]
Just would like to aplause you are working on this.

The most common fromat used in my country is DD-MM-YYYY by example.

**Reply by edwardtfn** [reply]
But in this case in the panel where each char space is precious, would you choose the year or the week day?

**Reply by rleongcs** [reply]
I do not believed this is an issue for what @Zaleo80 is asking for. Of course the week day must be in short format. You can already do that in your `Dev` branch easily. See below my sample.

What I use (%a, %-d %b %Y):
![IMG_1766](https://github.com/Blackymas/NSPanel_HA_Blueprint/assets/61628312/7f0bc6e1-8978-4f45-909a-b4fb04704b8b)

What I believed @Zaleo80 is looking for (%a, %d-%m-%Y):
![IMG_1767](https://github.com/Blackymas/NSPanel_HA_Blueprint/assets/61628312/ac7f1815-a79c-4119-9fcc-828237a2be2b)

**Reply by edwardtfn** [reply]
It looks nice. 🙂

**Reply by rleongcs** [reply]
@edwardtfn I guess that I only have you to thank for bringing about the long and short week day format. 🎉  Else I will always have to manually made some changes to the code myself to achieve it. 😆

**Reply by Zaleo80** [reply]
> What I believed @Zaleo80 is looking for (%a, %d-%m-%Y): ![IMG_1767](https://user-images.githubusercontent.com/61628312/241736892-ac7f1815-a79c-4119-9fcc-828237a2be2b.jpg)

Exactly this. Would be in line with the local standard. But really it is not that important. No hurry I wait till the custom formats are out of development.

**Comment by edwardtfn** [comment]
I'm closing this discussion as we have released v4.0 (#1057).
Please add a feature request for the missing things.

And thanks a lot for your support.
